src/workerd/jsg/filesystem.c++

```diff
@@ -0,0 +1,537 @@
+#include "filesystem.h"
+
+#include "buffersource.h"
+
+namespace workerd::jsg::fs {
+
+namespace {
+constexpr size_t kMaxReadLength = 32 * 1024;
+
+void validatePathWithNoSeparators(kj::StringPtr path) {
+  try {
+    auto parsed = kj::Path::parse(path);
+    JSG_REQUIRE(parsed.size() == 1, Error, "Invalid path: \"", path, "\"");
+  } catch (kj::Exception& e) {
+    JSG_FAIL_REQUIRE(Error, "Invalid path: \"", path, "\"");
+  }
+}
+
+template <bool Writable = false>
+class DirectoryBase final: public Directory {
+ public:
+  DirectoryBase() = default;
+  DirectoryBase(kj::HashMap<kj::String, kj::OneOf<kj::Own<File>, kj::Own<Directory>>> entries)
+      : entries(kj::mv(entries)) {
+    KJ_DASSERT(!Writable);
+  }
+
+  const Stat stat() const override {
+    return Stat{
+      .type = Node::Type::DIRECTORY,
+      .size = 0,
+      .lastModified = lastModified,
+      .writable = Writable,
+    };
+  }
+
+  size_t count(Lock& js, kj::Maybe<Type> typeFilter = kj::none) const override {
+    KJ_IF_SOME(type, typeFilter) {
+      size_t counter = 0;
+      forEach(js, [&counter](Lock&, kj::StringPtr name, const Stat stat) {
```

Comment:
forEach will execute it's callbacks if getDirectory() succeeds. but this implementation hides that fact. maybe we should rename this function, and similar ones.

```suggestion
      tryForEach(js, [&counter](Lock&, kj::StringPtr name, const Stat stat) {
```